kube-runtime/src/watcher.rs

```diff
@@ -467,18 +469,26 @@ where
             mut stream,
         } => match stream.next().await {
             Some(Ok(WatchEvent::Added(obj) | WatchEvent::Modified(obj))) => {
-                let resource_version = obj.resource_version().unwrap();
-                (Some(Ok(Event::Applied(obj))), State::Watching {
-                    resource_version,
-                    stream,
-                })
+                let resource_version = obj.resource_version().unwrap_or_default();
+                if resource_version.is_empty() {
+                    (Some(Err(Error::NoResourceVersion)), State::default())
```

Comment:
One minor observation here. Before this change in the watching stage, we would actually be somewhat resilient to missing resource versions here (where resource versions got transmitted as empty strings) because the resource version is only used in the next watch, and Kubernetes seem to accept empty strings in the watch params 🙃 

It is __technically possible__ that this change will lead to more re-lists because of some apiserver returning "" as resource versions. Still, that's kind of an impossibly defensive game to play against what is meant to be a stable api, so it's probably best to assume they are reasonably well behaved here, and raise bugs upstream otherwise.

Reply:
yes its theoretically possible, but as far as i'm aware, this should practically not happen since the resource version in the response is never blank. From https://kubernetes.io/docs/reference/using-api/api-concepts/#resourceversion-in-metadata:
> v1.meta/ListMeta - The metadata.resourceVersion of a resource collection (the response to a list) identifies the resource version at which the collection was constructed.